### ISSUE TITLE
Cosmos: Projecting out owned entities which own additional entity types retrieves the entire document

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#34067]
```c#
public virtual Task Project_owned_reference_navigation_which_owns_additional(bool async)
    => AssertQuery(
        async,
        ss => ss.Set<OwnedPerson>().OrderBy(o => o.Id).Select(p => p.PersonAddress));
```

Since PersonAddress owns further entity types (OwnedCountry), the expression inside the Select() contains IncludeExpression; this leads CosmosProjectionBindingExpressionVisitor into a specific path where we end up projecting the entire document. Compare how this works in relational - CosmosProjectionBindingExpressionVisitor is one place where Cosmos is still quite behind.

**Comment by roji** [comment]
Not actually closed by #34355

**Comment by ajcvickers** [comment]
@roji This is not limited to owned entities that have additional types. Projecting either an owned reference or collection results in bringing back the entire document. For example:

```C#
    await context
        .Set<Customer>()
        .AsNoTracking()
        .Select(e => e.Addresses)
        .ToListAsync();

    await context
        .Set<Customer>()
        .AsNoTracking()
        .Select(e => e.PhoneNumber)
        .ToListAsync();
```

```
info: 8/28/2024 12:33:56.712 CosmosEventId.ExecutingSqlQuery[30100] (Microsoft.EntityFrameworkCore.Database.Command) 
      Executing SQL query for container 'SomeDbContext' in partition 'None' [Parameters=[]]
      SELECT VALUE c
      FROM root c
info: 8/28/2024 12:33:56.719 CosmosEventId.ExecutedReadNext[30102] (Microsoft.EntityFrameworkCore.Database.Command) 
      Executed ReadNext (6.7595 ms, 2.25 RU) ActivityId='9616a1c2-9e2a-4858-a426-68e6001d2ac1', Container='SomeDbContext', Partition='None', Parameters=[]
      SELECT VALUE c
      FROM root c
```

```C#
public class Customer
{
    public Guid Id { get; set; }
    public required string Name { get; set; }
    public required string PrimaryCountry { get; set; }
    public required string Region { get; set; }
    public List<string>? Notes { get; set; }
    public Phone PhoneNumber { get; set; }
    public List<Address> Addresses { get; } = new();
}

public class Address
{
    public required string Street { get; set; }
    public required string City { get; set; }
    public required string Postcode { get; set; }
    public required string Country { get; set; }
}

public class Phone
{
    public required int CountryCode { get; set; }
    public required string Number { get; set; }
}
```

**Comment by roji** [comment]
@ajcvickers understood and thanks for checking... I think this makes this even more important to fix (in 10).

**Comment by roji** [comment]
Note: when we implement complex type support, this should cover properly projecting those out as well.

**Comment by JoasE** [comment]
Note: This will have an effect on distinct support as we currently can't apply DISTINCT on the query since the projection happens on the client
~2nd Note: Consider not selecting using VALUE, as it will filter out undefined (null) results in nested cases `Select(x => x.OptionalAssociate.OptionalNested)`, when OptionalAssociate is null~

**Comment by roji** [comment]
> 2nd Note: Consider not selecting using VALUE, as it will filter out undefined (null) results in nested cases Select(x => x.OptionalAssociate.OptionalNested), when OptionalAssociate is null

Note that `undefined` and `null` are very different things in Cosmos - a property can exist in the document and have a JSON `null` value, or it can be absent, in which case accessing it returns `undefined`. It's worth always keeping that distinction as the two values behave very differently.

In any case... The `undefined` filtering behavior is something we discussed at length back in the day, and older versions of the Cosmos provider actually did wrap projected scalar values in a JSON object, so that if they're `undefined` we'd still get a result. The decision we ended up making in this case was to align with the database's native way of doing things (so `VALUE`), even if it means that implicit filtering would happen, which could indeed be surprising.

**Comment by JoasE** [comment]
> Note: This will have an effect on distinct support as we currently can't apply DISTINCT on the query since the projection happens on the client

On this, also see: [#37577 (comment)](https://github.com/dotnet/efcore/pull/37577/changes#r2895441529)

**Comment by JoasE** [comment]
@roji @AndriySvyryd 
I'm (still) working on this, combined with #34567 (forgot to add a comment here also) and have run into something I'd like to discuss. It's related to the undefined filtering behaviour which was briefly discussed above, which causes conflicts when combined with projecting out owned types (at least with `AsNoTrackingWithIdentityResolution` or mappings)

> In any case... The undefined filtering behavior is something we discussed at length back in the day, and older versions of the Cosmos provider actually did wrap projected scalar values in a JSON object, so that if they're undefined we'd still get a result. The decision we ended up making in this case was to align with the database's native way of doing things (so VALUE), even if it means that implicit filtering would happen, which could indeed be surprising.

As stated in your comment above, EF will use `SELECT VALUE` when projecting out single values, causing undefined results to be filtered out implicitly. However IIUC, when projecting out an owned type as a single value (`context.Entities.Select(x => x.OptionalAssociate.OptionalNestedAssociate)`) EF needs to add new value(s) to the projection to retrieve the primary key of the (root) owner entity. I think these are really only actually needed for `AsNoTrackingWithIdentityResolution`, where EF needs them to perform identity resolution as owned entities are resolved by the identity of the (root) owner entity. Because there are now multiple values in the select, cosmos will not filter out undefined `OptionalNestedAssociate`'s.

On top of that, when projecting out a structural type as a dto/anon type (`context.Entities.Select(x => new{ x.OptionalAssociate.OptionalNestedAssociate.Id, x.OptionalAssociate.OptionalNestedAssociate.String })`), EF will translate to `SELECT c.OptionalAssociate.OptionalNestedAssociate.Id, c.OptionalAssociate.OptionalNestedAssociate.String` which also won't filter out undefined results (currently throws in 10.0). Whereas projecting out a Complex type specifically (`context.Entities.Select(x => x.OptionalAssociate.OptionalNestedAssociate)`) would filter that out, as it would be translated to `SELECT VALUE`

This also got me thinking that using a structural type projection vs using a mapping to a dto/anon type is currently already a bit inconsistent due to `SELECT VALUE` behaviour, since ef always projects single values as VALUE, selecting 2 or 1 property in an anon type has different behaviour in 10.0. However, this could be seen as a separate issue.
```csharp
var emptyList = await dbContext.Entities.AsNoTracking().Select(x => new { x.Associate!.NestedAssociate!.Id }).ToListAsync();
var throwsNullReference = await dbContext.Entities.AsNoTracking().Select(x => new { x.Associate!.NestedAssociate!.Id, x.Associate.NestedAssociate!.Name }).ToListAsync();
```

Unfortunately, using a `SELECT VALUE { "propN": "propN", ... }` doesn't help as it won't filter out undefined results either.
This all leads me to believe that `SELECT VALUE` isn't the right direction for EF, although I understand the argument for aligning with cosmos behaviour. Then again, you might have already considered these cases when discussing this back in the day.

Some things I have considered but don't think are viable:
- Ignoring results that don't define a projected property, this would mean that something the user or EF projected out is undefined and should have been filtered out
- Adding some form of filtering in the where (or even select clause?) to emulate SELECT VALUE behaviour when the user is projecting out a single value in the linq query

To summarize: I think the current select value behaviour, is already and, when implementing projecting out nested documents would cause unclear inconstancies in filtering behaviour between selecting out complex types, owned types and mappings thereof. 

My main question being: What is the desired behaviour in these cases for the completion of this issue and #34567

**Comment by roji** [comment]
@JoasE thanks for this analysis.

We indeed discussed similar cases back in EF 9 when redoing significant parts of the Cosmos query pipeline. Back then, of course, the discussion was restricted to projecting pure scalars, since the pipeline didn't (and still doesn't) support projecting out nested documents without the entire document.

So the cases you're outlining above are new, and I admit they're not great... Here are some thoughts:

1. For the first case, which IIUC is a problem only for AsNoTrackingWithIdentityResolution with owned entities; I generally think this isn't important enough. The thing to concentrate on going forward is IMHO complex types, where none of this can occur (complex types are always projected out as non-tracking and without any sort of identity). IMHO this also highlights the general weirdness of using the owned entity concept to map something that does not actually have an identity.
2. For entity vs. anonymous type projection, if we squint a bit, I think the behavior actually makes sense. Anonymous type projection means there's a `new` in the projection, so the result indeed should never be undefined (at least I hope my understanding of the current behavior is correct) - it only be missing certain properties inside (since they're undefined). Whereas projecting structural types via properties is inherently different, and if the projected structural type is truly absent from the document, then the general "filter out absent/undefined" makes sense (to me), even if I can see how it could be surprising. But I'm convinced it's any more surprising than the existing filtering-out behavior of scalars which we already have today.
3. I'm unaware of the behavioral difference you're describing between projecting an anonymous type with one property vs. two; that indeed sounds like a separate/unrelated bug. Can you open a separate issue with a repro?

The main thing I'm trying to do here, is to stick to the database semantics/behavior as much as possible, and to avoid EF imposing an external behavior which we believe is better. In other words, I'm not a fan of the Cosmos behavior around filtering out undefined (though I understand where it comes from - Cosmos returns JSON and there's no concept of undefined there); but that's the Cosmos behavior that we're surfacing. Another way of looking at this is that someone using Cosmos without EF will run into exactly the same behavior - it's just how Cosmos works - and I don't think it should generally be our job to change or "repair" database behavior, just to expose it.

But the above are just my own views - what are your thoughts on the above? Also, this is definitely something to be discussed with @AndriySvyryd.

**Comment by JoasE** [comment]
@roji That makes a lot of sense, thanks for explaining.

> The main thing I'm trying to do here, is to stick to the database semantics/behavior as much as possible, and to avoid EF imposing an external behavior which we believe is better. In other words, I'm not a fan of the Cosmos behavior around filtering out undefined (though I understand where it comes from - Cosmos returns JSON and there's no concept of undefined there); but that's the Cosmos behavior that we're surfacing. Another way of looking at this is that someone using Cosmos without EF will run into exactly the same behavior - it's just how Cosmos works - and I don't think it should generally be our job to change or "repair" database behavior, just to expose it.

Being the devils advocate, `SELECT VALUE` is **opt-in** behavior in the cosmos query langue. On the other hand, I did some testing and the (.NET) Cosmos SDK `ItemLinqQueryable` does also translate these cases to `SELECT VALUE`, and causes the same behavior for entity vs. anonymous type projection. So someone using the SDK would indeed run into the same and I think the `new` is never undefined argument makes sense there.

> For the first case, which IIUC is a problem only for AsNoTrackingWithIdentityResolution with owned entities; I generally think this isn't important enough. The thing to concentrate on going forward is IMHO complex types, where none of this can occur (complex types are always projected out as non-tracking and without any sort of identity). IMHO this also highlights the general weirdness of using the owned entity concept to map something that does not actually have an identity.

I had similar thoughts about this case as well, but was also wondering whether there might be other cases now (or very likely in the future?), where EF might need to project out 2 values where the user query is only selecting one, opening up the same question again.

Assuming we are keeping the `SELECT VALUE` behaviour, my main question on AsNoTrackingWithIdentityResolution is how I should move forward in the context of this issue. Are you implying we will just accept different filtering behavior for AsNoTrackingWithIdentityResolution for now? Maybe also good to note that relational currently always projects out the primary key of the owner entity, but I think I could make it so cosmos only projects it out when AsNoTrackingWithIdentityResolution is used. Or are you suggesting dropping support for AsNoTrackingWithIdentityResolution (edit: for owned entities) in cosmos, keeping in mind we will eventually drop support anyway?

**Comment by JoasE** [comment]
For context: in https://github.com/dotnet/efcore/issues/38298#issuecomment-4710527429 , it was discussed that `SELECT VALUE` will remain for now, but is subject for removal in the future.

@AndriySvyryd In this issue, I am still uncertain how to proceed with `AsNoTrackingWithIdentityResolution` considering it would cause inconsistent filtering behaviour.

See:
> Assuming we are keeping the SELECT VALUE behaviour, my main question on AsNoTrackingWithIdentityResolution is how I should move forward in the context of this issue. Are you implying we will just accept different filtering behavior for AsNoTrackingWithIdentityResolution for now? Maybe also good to note that relational currently always projects out the primary key of the owner entity, but I think I could make it so cosmos only projects it out when AsNoTrackingWithIdentityResolution is used. Or are you suggesting dropping support for AsNoTrackingWithIdentityResolution (edit: for owned entities) in cosmos, keeping in mind we will eventually drop support anyway?

**Comment by AndriySvyryd** [comment]
To get us to a place of some consistency I propose to apply this rule of thumb: "Any naked projection will be translated as `SELECT VALUE` and any top-level instantiation (DTO/anon/entity/complex) in the projection won't use `SELECT VALUE`".

`AsNoTrackingWithIdentityResolution` on a query that doesn't already project out the entity or its key should throw.

**Comment by AndriySvyryd** [comment]
@JoasE Just FYI - July 27 is the last day we can merge a big change like this one for v11
